### PR TITLE
[Proof of concept] Expand environment variables during preprocess

### PR DESCRIPTION
**Pull request by Kemyke**
Expanding environment vaariables in `UriDirectiveProcessor`.

### CODE REVIEW

File: tests/integration/build.ps1 [hunk 1]
```diff
@@ -50,6 +50,7 @@ Function GetCakeCoreCLRPath([string]$ScriptRoot)
 # PREPARATION
 #####################################################################
 
+$Env:CAKE_INTEGRATION_TEST_ROOT = "../.."
```
Comment (devlead) [inline tests/integration/build.ps1]: this should be after 
https://github.com/Kemyke/cake/blob/d8dd72d4ce3bdd6cd68bcf9c4db0655ff508b5e6/tests/integration/build.ps1#L174

where integration test variables are set,

Comment (devlead) [reply tests/integration/build.ps1]: Also we would need to do the same in `build.sh`

File: src/Cake.Core.Tests/Unit/Scripting/Processors/LoadDirectiveProcessorTests.cs [hunk 2]
```diff
@@ -69,6 +69,25 @@ public void Should_Process_Multiple_Script_References_Found_In_Source(string sou
             Assert.Equal("/Working/other.cake", result.Script.Includes[1].Path.FullPath);
         }
 
+        [Theory]
+        [InlineData("#l \"%CAKE_TEST_SCRIPT_PATH%/utils.cake\"")]
+        public void Should_Process_Environment_Variable_Found_In_Source(string source)
+        {
+            // Given
+            System.Environment.SetEnvironmentVariable("CAKE_TEST_SCRIPT_PATH", "/Working");
```
Comment (devlead) [inline src/Cake.Core.Tests/Unit/Scripting/Processors/LoadDirectiveProcessorTests.cs]: This fails on posix systems. Haven't checked why, this should only affect process, but could be something with how mono handles processes.

Comment (patriksvensson) [reply src/Cake.Core.Tests/Unit/Scripting/Processors/LoadDirectiveProcessorTests.cs]: To be honest I'm not a fan of setting environment variables like this in a test. It works but is more or less the same thing as a shared mutable state.

Comment (devlead) [reply src/Cake.Core.Tests/Unit/Scripting/Processors/LoadDirectiveProcessorTests.cs]: Agree doesn't "feel" ok. Don't know what would be better, picking existing known to exist? That doesn't feel 100% either.

**Comment by Kemyke** [comment]
I was thinking about to do this in `ScriptAnalyzer` but it looked quite risky to me. Maybe it is not a good idea to expand the env variables in all lines because we don't know the context. I mean it would replace variables in string, comments, etc. Furthermore this environment variable expansion is only important in the preprocessor directives. In the cake script you can easily access env vars with `EnvironmentVariable` function. But in the preprocess step it is not possible to use these functions or variables of course so my only idea was the environment variable to access various values during the preprocess step. 
That's why i implemented in `UriDirectiveProcessor` so the expansion happens only in URI-s. 

Or is it possible somehow to use variables in this directives? Like `$"#load {repoRoute}/utils.cake"` or `#addin "nuget:?package=Cake.Some.Nuget&version={someNugetVersion}"`. Would this be a better way?

**Comment by patriksvensson** [comment]
@Kemyke I accidentally merged this one to develop and I will have to revert it. Sorry for the inconvenience.

**Comment by patriksvensson** [comment]
@Kemyke As soon as the PR is reverted you should be able to rebase it against develop and submit it again.

Again, sorry about the inconvenience. First time this happened to me 😞

**Comment by devlead** [comment]
@Kemyke I've reverted so should just need to open a new pr.

**Comment by Kemyke** [comment]
No problem guys. I made a new [PR](https://github.com/cake-build/cake/pull/2023) and copied my thoughts and questions.

**Comment by Kemyke** [comment]
> Cause it doesn't change the source line, only the line passed to to the directives, if no directive accepts the line it'll just add the source line

Ok, now i understand and i agree it looks much better there.

**Pull request merged**